File: src/stream.ts
import type { RawCursorJson } from "./types.js";

export interface StreamState {
  lastTool: string | null;
  tokensSoFar: number;
  lastAssistant: string | null;
  filesTouched: string[];
  phase: string | null;
}

export function initStreamState(): StreamState {
  return {
    lastTool: null,
    tokensSoFar: 0,
    lastAssistant: null,
    filesTouched: [],
    phase: null,
  };
}

export interface ParsedLine {
  /** Set only for the terminal `result` event. */
  result?: RawCursorJson;
  /** Whether live state changed (a progress emit is worthwhile). */
  changed: boolean;
}

interface ToolCallShape {
  [key: string]: unknown;
  mcpToolCall?: { toolName?: string };
}

const PATH_ARG_KEYS = ["path", "filePath", "file_path", "target", "destination"];

function pathFromArgs(args: Record<string, unknown> | undefined): string | null {
  if (!args) return null;
  for (const key of PATH_ARG_KEYS) {
    const v = args[key];
    if (typeof v === "string" && v.length > 0) return v;
  }
  return null;
}

function extractTool(tc: ToolCallShape): {
  tool: string | null;
  path: string | null;
} {
  if (tc.mcpToolCall) {
    return { tool: tc.mcpToolCall.toolName ?? "mcp", path: null };
  }
  for (const key of Object.keys(tc)) {
    if (key.endsWith("ToolCall")) {
      const name = key.slice(0, -"ToolCall".length); // "shell", "edit", ...
      const args = (tc[key] as { args?: Record<string, unknown> } | undefined)
        ?.args;
      return { tool: name, path: pathFromArgs(args) };
    }
  }
  return { tool: null, path: null };
}

function extractAssistantText(message: unknown): string | null {
  const content = (message as { content?: unknown })?.content;
  if (!Array.isArray(content)) return null;
  const parts = content
    .filter(
      (b): b is { type: string; text: string } =>
        !!b && (b as { type?: string }).type === "text" &&
        typeof (b as { text?: unknown }).text === "string",
    )
    .map((b) => b.text);
  if (parts.length === 0) return null;
  return parts.join("");
}

/**
 * Parse one stream-json (NDJSON) line, mutating `state`.
 * Ignores blank / non-JSON lines and events with no `type`.
 */
export function parseLine(line: string, state: StreamState): ParsedLine {
  const trimmed = line.trim();
  if (!trimmed) return { changed: false };

  let ev: Record<string, unknown>;
  try {
    ev = JSON.parse(trimmed) as Record<string, unknown>;
  } catch {
    return { changed: false };
  }
  if (!ev || typeof ev.type !== "string") return { changed: false };

  let changed = false;

  // Any event carrying a numeric usage.outputTokens updates the live token count.
  const usage = ev.usage as { outputTokens?: unknown } | undefined;
  if (usage && typeof usage.outputTokens === "number") {
    state.tokensSoFar = Math.max(state.tokensSoFar, usage.outputTokens);
    changed = true;
  }

  switch (ev.type) {
    case "tool_call": {
      if (ev.subtype === "started" && ev.tool_call) {
        // A tool is now in flight — the caller's idle watchdog uses this to switch to the
        // wider toolIdleMs window, since a running build/test can go silent for a long time
        // without that being a hang.
        state.phase = "running_tool";
        changed = true;
        const { tool, path } = extractTool(ev.tool_call as ToolCallShape);
        if (tool) {
          state.lastTool = tool;
        }
        if (path && !state.filesTouched.includes(path)) {
          state.filesTouched.push(path);
        }
      }
      break;
    }
    case "assistant": {
      const text = extractAssistantText(ev.message);
      if (text) {
        state.lastAssistant = text.slice(0, 200);
        state.phase = "responding";
        changed = true;
      }
      break;
    }
    case "thinking": {
      if (ev.subtype === "delta" && typeof ev.text === "string") {
        state.phase = "thinking";
        changed = true;
      }
      break;
    }
    case "result":
      return { result: ev as RawCursorJson, changed: true };
  }

  return { changed };
}

File: src/stream.rs
use crate::types::RawCursorJson;
use serde_json::Value;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct StreamState {
    pub last_tool: Option<String>,
    pub tokens_so_far: f64,
    pub last_assistant: Option<String>,
    pub files_touched: Vec<String>,
    pub phase: Option<String>,
}

pub fn init_stream_state() -> StreamState {
    StreamState::default()
}

#[derive(Debug, Clone, Default)]
pub struct ParsedLine {
    pub result: Option<RawCursorJson>,
    pub changed: bool,
}

const PATH_ARG_KEYS: [&str; 5] = ["path", "filePath", "file_path", "target", "destination"];

fn path_from_args(args: Option<&Value>) -> Option<String> {
    let obj = args?.as_object()?;
    for key in PATH_ARG_KEYS {
        if let Some(v) = obj
            .get(key)
            .and_then(|x| x.as_str())
            .filter(|s| !s.is_empty())
        {
            return Some(v.to_string());
        }
    }
    None
}

fn extract_tool(tc: &Value) -> (Option<String>, Option<String>) {
    if let Some(mcp) = tc.get("mcpToolCall") {
        let tool = mcp
            .get("toolName")
            .and_then(|x| x.as_str())
            .unwrap_or("mcp")
            .to_string();
        return (Some(tool), None);
    }
    if let Some(obj) = tc.as_object() {
        for (key, val) in obj {
            if let Some(name) = key.strip_suffix("ToolCall") {
                let args = val.get("args");
                return (Some(name.to_string()), path_from_args(args));
            }
        }
    }
    (None, None)
}

fn extract_assistant_text(message: Option<&Value>) -> Option<String> {
    let content = message?.get("content")?.as_array()?;
    let parts: Vec<&str> = content
        .iter()
        .filter(|b| b.get("type").and_then(|t| t.as_str()) == Some("text"))
        .filter_map(|b| b.get("text").and_then(|t| t.as_str()))
        .collect();
    if parts.is_empty() {
        return None;
    }
    Some(parts.join(""))
}

pub fn parse_line(line: &str, state: &mut StreamState) -> ParsedLine {
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return ParsedLine {
            changed: false,
            result: None,
        };
    }
    let ev: Value = match serde_json::from_str(trimmed) {
        Ok(v) => v,
        Err(_) => {
            return ParsedLine {
                changed: false,
                result: None,
            };
        }
    };
    let ty = match ev.get("type").and_then(|t| t.as_str()) {
        Some(t) => t,
        None => {
            return ParsedLine {
                changed: false,
                result: None,
            };
        }
    };

    let mut changed = false;
    if let Some(n) = ev
        .pointer("/usage/outputTokens")
        .and_then(|x| x.as_f64())
        .filter(|n| n.is_finite())
    {
        state.tokens_so_far = state.tokens_so_far.max(n);
        changed = true;
    }

    match ty {
        "tool_call" => {
            if ev.get("subtype").and_then(|s| s.as_str()) == Some("started")
                && ev.get("tool_call").is_some()
            {
                state.phase = Some("running_tool".into());
                changed = true;
                let (tool, path) = extract_tool(ev.get("tool_call").unwrap());
                if let Some(t) = tool {
                    state.last_tool = Some(t);
                }
                if let Some(p) = path
                    && !state.files_touched.contains(&p)
                {
                    state.files_touched.push(p);
                }
            }
        }
        "assistant" => {
            if let Some(text) = extract_assistant_text(ev.get("message")) {
                let truncated: String = text.chars().take(200).collect();
                state.last_assistant = Some(truncated);
                state.phase = Some("responding".into());
                changed = true;
            }
        }
        "thinking" => {
            if ev.get("subtype").and_then(|s| s.as_str()) == Some("delta")
                && ev.get("text").and_then(|t| t.as_str()).is_some()
            {
                state.phase = Some("thinking".into());
                changed = true;
            }
        }
        "result" => {
            let raw: RawCursorJson = serde_json::from_value(ev).unwrap_or_default();
            return ParsedLine {
                result: Some(raw),
                changed: true,
            };
        }
        _ => {}
    }

    ParsedLine {
        result: None,
        changed,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn ignores_blank_and_non_json() {
        let mut s = init_stream_state();
        assert!(!parse_line("", &mut s).changed);
        assert!(!parse_line("   ", &mut s).changed);
        assert!(!parse_line("not json", &mut s).changed);
    }

    #[test]
    fn ignores_events_with_no_type() {
        let mut s = init_stream_state();
        assert!(!parse_line(&json!({"foo":1}).to_string(), &mut s).changed);
    }

    #[test]
    fn shell_tool_call_sets_last_tool() {
        let mut s = init_stream_state();
        parse_line(
            &json!({
                "type":"tool_call","subtype":"started",
                "tool_call":{"shellToolCall":{}}
            })
            .to_string(),
            &mut s,
        );
        assert_eq!(s.last_tool.as_deref(), Some("shell"));
        assert_eq!(s.phase.as_deref(), Some("running_tool"));
    }

    #[test]
    fn edit_tool_call_records_path() {
        let mut s = init_stream_state();
        parse_line(
            &json!({
                "type":"tool_call","subtype":"started",
                "tool_call":{"editToolCall":{"args":{"path":"src/a.ts"}}}
            })
            .to_string(),
            &mut s,
        );
        assert_eq!(s.last_tool.as_deref(), Some("edit"));
        assert_eq!(s.files_touched, ["src/a.ts"]);
    }

    #[test]
    fn write_delete_record_path_shaped_args() {
        let mut s = init_stream_state();
        parse_line(
            &json!({
                "type":"tool_call","subtype":"started",
                "tool_call":{"writeToolCall":{"args":{"filePath":"src/new.ts"}}}
            })
            .to_string(),
            &mut s,
        );
        assert_eq!(s.last_tool.as_deref(), Some("write"));
        assert_eq!(s.files_touched, ["src/new.ts"]);
        parse_line(
            &json!({
                "type":"tool_call","subtype":"started",
                "tool_call":{"deleteToolCall":{"args":{"path":"old.ts"}}}
            })
            .to_string(),
            &mut s,
        );
        assert_eq!(s.files_touched, ["src/new.ts", "old.ts"]);
    }

    #[test]
    fn mcp_tool_call_uses_tool_name() {
        let mut s = init_stream_state();
        parse_line(
            &json!({
                "type":"tool_call","subtype":"started",
                "tool_call":{"mcpToolCall":{"toolName":"search"}}
            })
            .to_string(),
            &mut s,
        );
        assert_eq!(s.last_tool.as_deref(), Some("search"));
        let mut s2 = init_stream_state();
        parse_line(
            &json!({
                "type":"tool_call","subtype":"started",
                "tool_call":{"mcpToolCall":{}}
            })
            .to_string(),
            &mut s2,
        );
        assert_eq!(s2.last_tool.as_deref(), Some("mcp"));
    }

    #[test]
    fn non_started_does_not_update() {
        let mut s = init_stream_state();
        parse_line(
            &json!({
                "type":"tool_call","subtype":"completed",
                "tool_call":{"shellToolCall":{}}
            })
            .to_string(),
            &mut s,
        );
        assert!(s.last_tool.is_none());
    }

    #[test]
    fn assistant_text_truncated_to_200() {
        let mut s = init_stream_state();
        let long = "x".repeat(300);
        parse_line(
            &json!({
                "type":"assistant",
                "message":{"content":[{"type":"text","text": long}]}
            })
            .to_string(),
            &mut s,
        );
        assert_eq!(s.last_assistant.as_ref().map(|t| t.len()), Some(200));
        assert_eq!(s.phase.as_deref(), Some("responding"));
    }

    #[test]
    fn thinking_delta_sets_phase() {
        let mut s = init_stream_state();
        parse_line(
            &json!({"type":"thinking","subtype":"delta","text":"hmm"}).to_string(),
            &mut s,
        );
        assert_eq!(s.phase.as_deref(), Some("thinking"));
    }

    #[test]
    fn usage_output_tokens_updates() {
        let mut s = init_stream_state();
        parse_line(
            &json!({"type":"assistant","usage":{"outputTokens":42}}).to_string(),
            &mut s,
        );
        assert_eq!(s.tokens_so_far, 42.0);
    }

    #[test]
    fn tokens_so_far_monotonic() {
        let mut s = init_stream_state();
        parse_line(
            &json!({"type":"assistant","usage":{"outputTokens":100}}).to_string(),
            &mut s,
        );
        parse_line(
            &json!({"type":"assistant","usage":{"outputTokens":40}}).to_string(),
            &mut s,
        );
        assert_eq!(s.tokens_so_far, 100.0);
    }

    #[test]
    fn result_event_returns_raw() {
        let mut s = init_stream_state();
        let r = parse_line(
            &json!({
                "type":"result","subtype":"success","is_error":false,
                "result":"final","session_id":"sid","duration_ms":99,
                "usage":{"outputTokens":10}
            })
            .to_string(),
            &mut s,
        );
        assert!(r.result.is_some());
        assert_eq!(r.result.as_ref().unwrap().result.as_deref(), Some("final"));
        assert_eq!(
            r.result.as_ref().unwrap().session_id.as_deref(),
            Some("sid")
        );
        assert_eq!(s.tokens_so_far, 10.0);
    }
}
